Build failure on macOS

```console
====================[ Build | Client.test | Debug ]=============================
"/Users/georgiy.lebedev/Library/Application Support/JetBrains/Toolbox/apps/CLion/ch-2/231.8109.222/CLion.app/Contents/bin/cmake/mac/bin/cmake" --build /Users/georgiy.lebedev/Work/tntcxx/build-debug --target Client.test -j 6
[1/2] Building CXX object CMakeFiles/Client.test.dir/test/ClientTest.cpp.o
FAILED: CMakeFiles/Client.test.dir/test/ClientTest.cpp.o 
ccache /usr/local/opt/llvm/bin/clang++  -I/Users/georgiy.lebedev/Work/tntcxx/. -I/Users/georgiy.lebedev/Work/tntcxx/./src -I/Users/georgiy.lebedev/Work/tntcxx/./third_party/libev -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -fcolor-diagnostics -Wall -Wextra -Werror -std=gnu++17 -MD -MT CMakeFiles/Client.test.dir/test/ClientTest.cpp.o -MF CMakeFiles/Client.test.dir/test/ClientTest.cpp.o.d -o CMakeFiles/Client.test.dir/test/ClientTest.cpp.o -c /Users/georgiy.lebedev/Work/tntcxx/test/ClientTest.cpp
In file included from /Users/georgiy.lebedev/Work/tntcxx/test/ClientTest.cpp:36:
In file included from /Users/georgiy.lebedev/Work/tntcxx/test/../src/Client/Connector.hpp:37:
In file included from /Users/georgiy.lebedev/Work/tntcxx/./src/../src/Client/UnixPlainStream.hpp:33:
/Users/georgiy.lebedev/Work/tntcxx/./src/../src/Client/UnixStream.hpp:171:28: error: use of undeclared identifier 'SOCK_NONBLOCK'
                              inf.ai_socktype | SOCK_NONBLOCK | SOCK_CLOEXEC,
                                                ^
/Users/georgiy.lebedev/Work/tntcxx/./src/../src/Client/UnixStream.hpp:171:44: error: use of undeclared identifier 'SOCK_CLOEXEC'
                              inf.ai_socktype | SOCK_NONBLOCK | SOCK_CLOEXEC,
                                                                ^
2 errors generated.
ninja: build stopped: subcommand failed.
```

--- a/examples/Simple.cpp
+++ b/examples/Simple.cpp
@@ -207,7 +207,7 @@ main()
 	printResponse<Buf_t>(conn, *response);
 	//doclabel11-2
 	/* Let's wait for both futures at once. */
-	std::vector<rid_t> futures;
+	std::vector<rid_t> futures(2);
 	futures[0] = replace;
 	futures[1] = select;
 	/* No specified timeout means that we poll futures until they are ready.*/

--- a/src/Buffer/Buffer.hpp
+++ b/src/Buffer/Buffer.hpp
@@ -158,10 +158,15 @@ public:
 	};
 	template <bool LIGHT>
 	class iterator_common
-		: public std::iterator<std::input_iterator_tag, char>,
-		  public std::conditional_t<LIGHT, light_base, SingleLink<iterator_common<LIGHT>>>
+		: public std::conditional_t<LIGHT, light_base, SingleLink<iterator_common<LIGHT>>>
 	{
 	public:
+		using iterator_category = std::input_iterator_tag;
+		using value_type = char;
+		using difference_type = std::ptrdiff_t;
+		using pointer = char *;
+		using reference = char &;
+
 		using Base_t = std::conditional_t<LIGHT, light_base, SingleLink<iterator_common<LIGHT>>>;
 		USING_LIST_LINK_METHODS(Base_t);
 

--- a/src/Client/ResponseReader.hpp
+++ b/src/Client/ResponseReader.hpp
@@ -150,9 +150,15 @@ template <class BUFFER>
 struct TupleReader : mpp::ReaderTemplate<BUFFER> {
 
 	TupleReader(mpp::Dec<BUFFER>& d, Data<BUFFER>& dt) : dec(d), data(dt) {}
-	static constexpr mpp::Family VALID_TYPES = mpp::MP_ARR | mpp::MP_UINT |
-		mpp::MP_INT | mpp::MP_BOOL | mpp::MP_DBL | mpp::MP_STR | mpp::MP_NIL;
+	static constexpr mpp::Family VALID_TYPES = mpp::MP_ARR | mpp::MP_MAP |
+		mpp::MP_UINT | mpp::MP_INT | mpp::MP_BOOL | mpp::MP_DBL |
+		mpp::MP_STR | mpp::MP_NIL;
 	void Value(iterator_t<BUFFER>& arg, mpp::compact::Family, mpp::ArrValue u)
+	{
+		data.tuples.emplace_back(arg, u.size);
+		dec.Skip();
+	}
+	void Value(iterator_t<BUFFER>& arg, mpp::compact::Family, mpp::MapValue u)
 	{
 		data.tuples.emplace_back(arg, u.size);
 		dec.Skip();

--- a/src/Client/ResponseReader.hpp
+++ b/src/Client/ResponseReader.hpp
@@ -151,7 +151,7 @@ struct TupleReader : mpp::ReaderTemplate<BUFFER> {
 
 	TupleReader(mpp::Dec<BUFFER>& d, Data<BUFFER>& dt) : dec(d), data(dt) {}
 	static constexpr mpp::Family VALID_TYPES = mpp::MP_ARR | mpp::MP_UINT |
-		mpp::MP_INT | mpp::MP_BOOL | mpp::MP_DBL | mpp::MP_STR; //| mpp::MP_NIL;
+		mpp::MP_INT | mpp::MP_BOOL | mpp::MP_DBL | mpp::MP_STR | mpp::MP_NIL;
 	void Value(iterator_t<BUFFER>& arg, mpp::compact::Family, mpp::ArrValue u)
 	{
 		data.tuples.emplace_back(arg, u.size);

--- a/src/Client/UnixStream.hpp
+++ b/src/Client/UnixStream.hpp
@@ -30,6 +30,7 @@
 */
 #pragma once
 
+#include <fcntl.h>
 #include <poll.h>
 #include <unistd.h>
 #include <sys/socket.h>
@@ -167,10 +168,18 @@ UnixStream::connect(const ConnectOptions &opts_arg)
 			      addr_info.last_error());
 	int socket_errno = 0, connect_errno = 0;
 	for (auto &inf: addr_info) {
-		fd = ::socket(inf.ai_family,
-			      inf.ai_socktype | SOCK_NONBLOCK | SOCK_CLOEXEC,
-			      inf.ai_protocol);
+		fd = ::socket(inf.ai_family, inf.ai_socktype, inf.ai_protocol);
 		if (fd < 0) {
+			socket_errno = errno;
+			continue;
+		}
+		if (fcntl(fd, F_SETFD, FD_CLOEXEC) < 0) {
+			::close(fd);
+			socket_errno = errno;
+			continue;
+		}
+		if (fcntl(fd, F_SETFL, O_NONBLOCK) < 0) {
+			::close(fd);
 			socket_errno = errno;
 			continue;
 		}

--- a/src/Utils/AddrInfo.hpp
+++ b/src/Utils/AddrInfo.hpp
@@ -60,11 +60,16 @@ private:
 	int rc = 0;
 };
 
-class AddrInfo::iterator :
-	std::iterator<std::input_iterator_tag, const struct addrinfo>
+class AddrInfo::iterator
 {
+public:
+    using iterator_category = std::input_iterator_tag;
+    using value_type = const struct addrinfo;
+    using difference_type = std::ptrdiff_t;
+    using pointer = const struct addrinfo *;
+    using reference = const struct addrinfo &;
 private:
-	using T = const struct addrinfo;
+	using T = value_type;
 public:
 	T &operator*() const { return *info; }
 	T *operator->() const { return info; }

--- a/test/ClientTest.cpp
+++ b/test/ClientTest.cpp
@@ -546,6 +546,7 @@ single_conn_call(Connector<BUFFER, NetProvider> &client)
 	const static char *return_select  = "remote_select";
 	const static char *return_uint    = "remote_uint";
 	const static char *return_multi   = "remote_multi";
+	const static char *return_nil     = "remote_nil";
 
 	Connection<Buf_t, NetProvider> conn(client);
 	int rc = test_connect(client, conn, localhost, port);
@@ -613,6 +614,13 @@ single_conn_call(Connector<BUFFER, NetProvider> &client)
 	fail_unless(response != std::nullopt);
 	fail_unless(response->body.error_stack != std::nullopt);
 	printResponse<BUFFER, NetProvider>(conn, *response);
+
+	TEST_CASE("call remote_nil");
+	rid_t f9 = conn.call(return_nil, std::make_tuple());
+	client.wait(conn, f9, WAIT_TIMEOUT);
+	fail_unless(conn.futureIsReady(f9));
+	response = conn.getResponse(f9);
+	printResponse<BUFFER, NetProvider>(conn, *response, MULTI_RETURN);
 
 	client.close(conn);
 }

--- a/test/ClientTest.cpp
+++ b/test/ClientTest.cpp
@@ -547,6 +547,7 @@ single_conn_call(Connector<BUFFER, NetProvider> &client)
 	const static char *return_uint    = "remote_uint";
 	const static char *return_multi   = "remote_multi";
 	const static char *return_nil     = "remote_nil";
+	const static char *return_map     = "remote_map";
 
 	Connection<Buf_t, NetProvider> conn(client);
 	int rc = test_connect(client, conn, localhost, port);
@@ -620,6 +621,13 @@ single_conn_call(Connector<BUFFER, NetProvider> &client)
 	client.wait(conn, f9, WAIT_TIMEOUT);
 	fail_unless(conn.futureIsReady(f9));
 	response = conn.getResponse(f9);
+	printResponse<BUFFER, NetProvider>(conn, *response, MULTI_RETURN);
+
+	TEST_CASE("call remote_map");
+	rid_t f10 = conn.call(return_map, std::make_tuple());
+	client.wait(conn, f10, WAIT_TIMEOUT);
+	fail_unless(conn.futureIsReady(f10));
+	response = conn.getResponse(f10);
 	printResponse<BUFFER, NetProvider>(conn, *response, MULTI_RETURN);
 
 	client.close(conn);

--- a/test/Utils/TupleReader.hpp
+++ b/test/Utils/TupleReader.hpp
@@ -39,13 +39,14 @@ struct UserTuple {
 	uint64_t field1;
 	std::string field2;
 	double field3;
+	void *field4 = reinterpret_cast<void *>(0xDEADBEEF);
 };
 
 std::ostream&
 operator<<(std::ostream& strm, const UserTuple &t)
 {
 	return strm << "Tuple: field1=" << t.field1 << " field2=" << t.field2 <<
-		    " field3=" << t.field3;
+		    " field3=" << t.field3 << " field4=" << t.field4;
 }
 
 using Buf_t = tnt::Buffer<16 * 1024>;
@@ -54,7 +55,8 @@ template <class BUFFER>
 struct TupleValueReader : mpp::DefaultErrorHandler {
 	using BufIter_t = typename BUFFER::iterator;
 	explicit TupleValueReader(mpp::Dec<BUFFER>& d, UserTuple& t) : dec(d), tuple(t) {}
-	static constexpr mpp::Family VALID_TYPES = mpp::MP_UINT | mpp::MP_STR | mpp::MP_DBL;
+	static constexpr mpp::Family VALID_TYPES = mpp::MP_UINT | mpp::MP_STR |
+		mpp::MP_DBL | mpp::MP_NIL;
 	template <class T>
 	void Value(const BufIter_t&, mpp::compact::Family, T v)
 	{
@@ -73,6 +75,10 @@ struct TupleValueReader : mpp::DefaultErrorHandler {
 			++tmp;
 			--v.size;
 		}
+	}
+	void Value(const BufIter_t&, mpp::compact::Family, std::nullptr_t)
+	{
+		tuple.field4 = nullptr;
 	}
 	void WrongType(mpp::Family expected, mpp::Family got)
 	{

--- a/test/Utils/TupleReader.hpp
+++ b/test/Utils/TupleReader.hpp
@@ -40,23 +40,53 @@ struct UserTuple {
 	std::string field2;
 	double field3;
 	void *field4 = reinterpret_cast<void *>(0xDEADBEEF);
+	std::pair<std::string, uint64_t> field5;
 };
 
 std::ostream&
 operator<<(std::ostream& strm, const UserTuple &t)
 {
 	return strm << "Tuple: field1=" << t.field1 << " field2=" << t.field2 <<
-		    " field3=" << t.field3 << " field4=" << t.field4;
+		    " field3=" << t.field3 << " field4=" << t.field4 <<
+		    " field5.first=" << t.field5.first <<
+		    " field5.second=" << t.field5.second;
 }
 
 using Buf_t = tnt::Buffer<16 * 1024>;
+
+template <class BUFFER>
+struct KeyValueReader : mpp::DefaultErrorHandler {
+	using BufIter_t = typename BUFFER::iterator;
+	explicit KeyValueReader(mpp::Dec<BUFFER>& d, std::pair<std::string, uint64_t>& kv) : dec(d), key_value(kv) {}
+	static constexpr mpp::Family VALID_TYPES = mpp::MP_STR;
+	void Value(BufIter_t& itr, mpp::compact::Family, mpp::StrValue v)
+	{
+		BufIter_t tmp = itr;
+		tmp += v.offset;
+		std::string &dst = key_value.first;
+		while (v.size) {
+			dst.push_back(*tmp);
+			++tmp;
+			--v.size;
+		}
+		dec.SetReader(true, mpp::SimpleReader<BUFFER, mpp::MP_UINT, uint64_t>{key_value.second});
+	}
+	void WrongType(mpp::Family expected, mpp::Family got)
+	{
+		std::cout << "expected type is " << expected << " but got " <<
+			got << std::endl;
+	}
+	BufIter_t* StoreEndIterator() { return nullptr; }
+	mpp::Dec<BUFFER>& dec;
+	std::pair<std::string, uint64_t>& key_value;
+};
 
 template <class BUFFER>
 struct TupleValueReader : mpp::DefaultErrorHandler {
 	using BufIter_t = typename BUFFER::iterator;
 	explicit TupleValueReader(mpp::Dec<BUFFER>& d, UserTuple& t) : dec(d), tuple(t) {}
 	static constexpr mpp::Family VALID_TYPES = mpp::MP_UINT | mpp::MP_STR |
-		mpp::MP_DBL | mpp::MP_NIL;
+		mpp::MP_DBL | mpp::MP_NIL | mpp::MP_MAP;
 	template <class T>
 	void Value(const BufIter_t&, mpp::compact::Family, T v)
 	{
@@ -79,6 +109,11 @@ struct TupleValueReader : mpp::DefaultErrorHandler {
 	void Value(const BufIter_t&, mpp::compact::Family, std::nullptr_t)
 	{
 		tuple.field4 = nullptr;
+	}
+	void Value(const BufIter_t&, mpp::compact::Family, mpp::MapValue v)
+	{
+		assert(v.size == 1);
+		dec.SetReader(false, KeyValueReader{dec, tuple.field5});
 	}
 	void WrongType(mpp::Family expected, mpp::Family got)
 	{

--- a/test/cfg.lua
+++ b/test/cfg.lua
@@ -26,6 +26,10 @@ function remote_nil()
     return nil
 end
 
+function remote_map()
+    return {key = 777}
+end
+
 function bench_func(...)
     return {...}
 end

--- a/test/cfg.lua
+++ b/test/cfg.lua
@@ -22,6 +22,10 @@ function remote_multi()
     return 'Hello', 1, 6.66
 end
 
+function remote_nil()
+    return nil
+end
+
 function bench_func(...)
     return {...}
 end

